Bluetooth: Mesh: add mbedtls as crypto library for mesh in Kconfig

alxelax opened the pull request:
Add mbedtls library as an option for crypto algorithm for
BLE mesh in Kconfig, cmake and configuration files.
Make ability to choose between mbedtls and tinycrypt for BLE mesh.

Signed-off-by: Aleksandr Khromykh <Aleksandr.Khromykh@nordicsemi.no>

alxelax commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,639 @@
+/*
```
files crypto_tc and crypto_mtls are mostly the same. I created the task to make security.c with common part and keep the only difference in crypto_tc and crypto_mtls

alxelax commented:
PSA crypto API doc (v1.0) https://armmbed.github.io/mbed-crypto/1.0.1/html/index.html
TF-M supports the same API. These changes should be compiled with TF-M without refactoring.

PavelVPV commented on modules/mbedtls/Kconfig:
```diff
@@ -50,6 +50,7 @@ config CUSTOM_MBEDTLS_CFG_FILE
 config MBEDTLS_CFG_FILE
 	string "mbed TLS configuration file" if CUSTOM_MBEDTLS_CFG_FILE
 	depends on MBEDTLS_BUILTIN
+	default "config-psa.h" if BT_MESH_USES_MBEDTLS_PSA
```
Isn't this configured through `CONFIG_MBEDTLS_USER_CONFIG_FILE`? Like in psa_crypto sample.

alxelax replied:
Usage of `CONFIG_MBEDTLS_USER_CONFIG_FILE` requires the presence of` Kconfig.tls-generic` and carries all unnecessary settings that are present for tls functionality. As an alternative, it is possible to use `CONFIG_CUSTOM_MBEDTLS_CFG_FILE` but need to understand that it will require adding in each prj.conf for mesh apps

alxelax replied:
not necessary anymore

PavelVPV commented on modules/mbedtls/configs/config-psa.h:
```diff
@@ -0,0 +1,27 @@
+
+#define MBEDTLS_AES_ROM_TABLES
+
+#include "mbedtls/config_psa.h"
+#include "mbedtls/check_config.h"
```
Isn't this included through `include/psa/crypto.h`? As per my understanding, there shouldn't be any include in the config header, only the configuration.

alxelax replied:
There is specific logic in these two files in the configuration header that should be used.
The file keeps the configuration definitions that the customer provides. There are a lot of them and most of them have explicit and implicit dependencies on each other.
There are, as I call them, "root" definitions like for example `MBEDTLS_CMAC_C`.
Header `config_psa.h` based on "root" definitions declares a lot of "subroot" definitions required to enable wished functionality.
Header `check_config.h` performs checking dependencies and generates compile time error if there is a lack of them or there are controversial settings.

I spent a quite long time figuring out which options should be used because including "subroot" options caused a lack of them and a request to add functionality that we really do not need.

I guess the current including as how it was intended to use by design.

PavelVPV replied:
From what I see, "psa/crypto.h" includes `psa/crypto_platform.h`. `psa/crypto_platform.h` includes `mbedtls/build_info.h`. `mbedtls/build_info.h` includes `MBEDTLS_USER_CONFIG_FILE` and these 2 headers. So that user application just defines `MBEDTLS_USER_CONFIG_FILE` and includes `psa/crypto.h`. This is how I guess it is designed. 

The other question I guess is how to pass some configuration that is mandatory for mesh though mesh is not an end application.

I feel that `Kconfig.tls-generic` and `config-tls-generic.h` (https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/mbedtls/configs/config-tls-generic.h) were added to configure mbedtls through Kconfig. So I guess what we actually have to do is to select options from Kconfig.tls-generic. If there are missing options, we need to add those into `Kconfig.tls-generic` and `config-tls-generic.h`. Thus we don't need extra header file, we don't break mbedtls configuration when something else other than mesh uses mbedtls and defines its own config file, which is included normally through `psa/crypto.h` (or somewhere else), a user can still add its config file to an application without breaking our.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -17,6 +13,22 @@ menuconfig BT_MESH
 
 if BT_MESH
 
+config BT_MESH_USES_TINYCRYPT
```
This should be a choice between either Tinycrypt or PSA. Which one should be selected by default I'm not sure. mbedtls give sbigger footprint, but Tinycrypt should as per my understanding by deprecated.

PavelVPV commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,639 @@
+		err = -ERANGE;
+	}
+
+	end:
```
extra indent.

PavelVPV commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,639 @@
+//		return -EIO;
+//	}
+
+	return 0;
```
Minor. Probably better to just return error here to not keep commented code in commits.

PavelVPV commented on tests/bluetooth/mesh_shell/prj.conf:
```diff
@@ -16,10 +16,12 @@ CONFIG_FLASH_MAP=y
 CONFIG_SETTINGS=y
 
+CONFIG_BT_MESH_USES_MBEDTLS_PSA=y
+CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE=1024
```
Shouldn't `MINIMAL_LIBC_MALLOC_ARENA_SIZE` config be set in `BT_MESH_USES_MBEDTLS_PSA`?

alxelax replied:
Hm. I tried to find how it is possible to do but seems Kconfig syntax doesn't allow to do something like:
```
config BT_MESH_USES_MBEDTLS_PSA
	bool "Use MBEDTLS PSA"
	select MBEDTLS
	select BT_HOST_CCM
        select MINIMAL_LIBC_MALLOC_ARENA_SIZE=1024
	help
	  Use MBEDTLS library to perform crypto operations.
```
Probably I misunderstood your idea. Could you elaborate this?

PavelVPV replied:
Right, this can only be done in the option itself.

But, actually, I can't seem to find where malloc is used in mbedtls. Why do we need this option?

alxelax replied:
This is the result of provisioning without CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE. I'm not sure how they are related to mbedTLS
```
uart:~$ mesh init
Mesh shell initialized
uart:~$ mesh pb-gatt on
PB-GATT enabled
[00:00:14.211,273] <inf> bt_mesh_prov_device: Device UUID: 00000000-0000-0000-0000-00000000dddd
Provisioning link opened on PB-GATT
[00:01:19.567,352] <err> os: CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE is 0
[00:01:19.567,382] <err> bt_mesh_prov_device: Unable to generate confirmation salt
Provisioning link closed on PB-GATT
```

PavelVPV replied:
Alright, I found this. If I understand correctly, with the current configuration, imported keys are kept in RAM, but ideally should be stored in ITS. With TF-M keys will be stored in ITS. I guess we can keep it for now until we get more understanding of how this should work for devices without TF-M support.

alxelax replied:
Yes, the current solution with keys is temporary. After normal implementation the keys should be imported once when Mesh stack gets them. However, it is not possible to refuse fully not persistent key usage i.e. during Diffie-Hellman shared secret generation.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -17,6 +13,29 @@ menuconfig BT_MESH
+	  Use TinyCrypt library to perform crypto operations.
+
+config BT_MESH_USES_MBEDTLS_PSA
+	bool "Use MBEDTLS PSA"
```
```suggestion
	bool "Use mbed TLS PSA"
```

PavelVPV commented on modules/Kconfig:
```diff
@@ -35,6 +35,7 @@ source "modules/Kconfig.vega"
 source "modules/Kconfig.xtensa"
 source "modules/zcbor/Kconfig"
 source "modules/Kconfig.mcuboot_bootutil"
+source "modules/mbedtls/Kconfig"
```
I'm not sure this is needed. mbedTLS configuration is available without this line.

PavelVPV commented on modules/mbedtls/CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ if(CONFIG_MBEDTLS_BUILTIN)
 
   target_include_directories(mbedTLS INTERFACE
     ${ZEPHYR_CURRENT_MODULE_DIR}/include
+    ${ZEPHYR_CURRENT_MODULE_DIR}/library
```
I guess this is because `psa_crypto_cipher.h` is included incorrectly: 
https://github.com/zephyrproject-rtos/mbedtls/blob/7fed49c9b9f983ad6416986661ef637459723bcb/library/psa_crypto_cipher.c#L25

I guess instead of adding `library` as include directory, it should be fixed in mbedtls repo. At least this is an internal header which obviously should be included with double quotes and in 2 other places where this file is included, it is included with double quotes.

alxelax replied:
I suggest to leave it `as is` in PR and to create a separate task to solve this in the right manner. It will require at least 2 PRs (one with west mbtle link update).

alxelax replied:
https://github.com/zephyrproject-rtos/mbedtls/pull/37   PR in mbedtls

alxelax committed:
Bluetooth: Mesh: split crypto common part in separate file

Most mesh crypto source code is independent of
the used crypto libraries. It is worth splitting it into
a separate file and to keep in the library-specific files
the only specific source code.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,643 @@
+{
+	return -ENOSYS;
+}
+
```
for some reason after migration on Zephyr's configuration mbedTLS mandated to have implementation of this function. 
Another option is to have internal entropy that is based on sha256 or sha512. I wasn't able to ron mbtls with entropy_c option. 
Probably it requires more config options that the checker cannot catch?

PavelVPV replied:
`MBEDTLS_TLS_VERSION_1_2` and `MBEDTLS_KEY_EXCHANGE_RSA_ENABLED` need to be disabled explicitly. They are enabled by default and pool `mbedtls_psa_external_get_random` through `psa_import_key`.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -17,6 +13,34 @@ menuconfig BT_MESH
+	select MBEDTLS_AES_ROM_TABLES
+	select BT_HOST_CCM
+	help
+	  Use MBEDTLS library to perform crypto operations.
```
Nitpicking:
```suggestion
	  Use mbed TLS library to perform crypto operations.
```

PavelVPV commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,644 @@
+	psa_set_key_type(&attributes, PSA_KEY_TYPE_AES);
+	psa_set_key_bits(&attributes, 128);
+
+	if (psa_import_key(&attributes, key, 16, &key_id) != PSA_SUCCESS) {
```
I guess this is an intermediate implementation as ideally keys should be imported only once?

alxelax replied:
yes, I added this for now. To work with provided key_id CDB and key representation within of Mesh and Host should be changed. I guess this is the most complex part of PSA support.

PavelVPV reviewed (approved):
This looks OK for now, but I feel we have some technical debt here:
- Should we do anything we key storage? Currently keys are stored in RAM, but I guess should be kept in ITS, or this doesn't matter for devices that doesn't support TF-M?
- Should we do anything with entropy source?

alxelax commented:
> This looks OK for now, but I feel we have some technical debt here:
> 
>     * Should we do anything we key storage? Currently keys are stored in RAM, but I guess should be kept in ITS, or this doesn't matter for devices that doesn't support TF-M?

Yes, this is separate huge task for reworking CDB and key importing\exporting in places when mesh provides\requests them.
Additionally, there should be abstract key type that changes representation by configuration compile time. 

> 
>     * Should we do anything with entropy source?

No for now. It seems well protected getting real random from radio. There is the only one thing. Probably it should be moved to the Host instead of Mesh.

alxelax committed:
Bluetooth: Mesh: add mbedtls as crypto library for mesh in Kconfig

Add mbedtls library as an option for crypto algorithm for
BLE mesh in Kconfig, cmake and configuration files.
Make ability to choose between mbedtls and tinycrypt for BLE mesh.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
mbedtls: add PSA build as config option

Add ability to build mbedtls with PSA API into
mbedtls build configuration files within Zephyr.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax commented:
Split commit into two commits one for mbedtls one for mesh

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,644 @@
+	int err;
+
+	err = bt_mesh_aes_cmac_one(salt, ikm, ikm_len, okm);
+	if (err < 0) {
```
```suggestion
	if (err) {
```

alxelax replied:
This is the initial crypto implementation that I moved to another file. I really do not want to do any bug fixing or\and refactoring within my task. I want to have here only changes related to mbedtls and PSA support.
If we decide to roll out the changes the source code shouldn't lose quality.

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,644 @@
+		return err;
+	}
+
+	err = bt_mesh_aes_cmac_one(tmp, n, 16, t);
```
Why not sizeof(n) instead of 16? More of them bellow.

alxelax replied:
because this is 16 bytes crypto. No possible other digits in the crypto mesh toolchain. Any case this is an initial implementation. I did refactoring related to hardcoded sizes in Mesh 1.1 repo for EHA. No reason to do it here since it will be replaced after merge Mesh 1.1

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,644 @@
+		return err;
+	}
+
+	for (i = 0; i < 6; i++) {
```
```suggestion
	for (int i = 0; i < 6; i++) {
```

alxelax replied:
the same reason regarding refactoring here as well.

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,644 @@
+
+	BT_DBG("Nonce %s", bt_hex(nonce, 13));
+
+	err = bt_ccm_encrypt(key, nonce, &buf->data[7], buf->len - 7, NULL, 0,
```
What will happen if buf->len is less than 7? Is that possible at all?

alxelax replied:
If buff is less than 7 then this is bug in mesh implementation and has to be fixed in mesh. This is a permanent network header size that is always present and used for obfuscation.
Need to consider this is aes ccm for mesh not for any purpose.
`3.4.4 Network PDU`

Balaklaka commented on subsys/bluetooth/mesh/crypto_mtls.c:
```diff
@@ -0,0 +1,644 @@
+int bt_mesh_prov_decrypt(const uint8_t key[16], uint8_t nonce[13],
+			 const uint8_t data[25 + 8], uint8_t out[25])
+{
+	return bt_ccm_decrypt(key, nonce, data, 25, NULL, 0, out, 8);
```
This is for both decrypt/encrypt, is the data size of 25 and out size of 8 correct when the arrays are 33 and 25?

alxelax replied:
yep. this is authentication calculation. 8 bytes is mic in Provisioning Confirmation frame.
`5.4.1.8 Provisioning Data`

alxelax commented:
@PavelVPV, @Balaklaka, I added commit from the task about splitting mesh module into general and specific parts here. 
I see that there was a misunderstanding and @Balaklaka started reviewing the existing crypto code.  
After this commit PR shows only related to PSA changes. 

Please review.

PavelVPV commented:
Btw, will you merge it on your `main` branch or maybe better to use another branch?

alxelax commented:
> Btw, will you merge it on your `main` branch or maybe better to use another branch?

I'm going to use my `main`.
